File: tools/test_excel_quad_parser.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

function splitCombinedBatches(raw) {
  if (!raw) return [];
  const s = String(raw).trim();
  if (s === '' || s.toLowerCase() === 'none' || s.toLowerCase() === 'date' || s.toLowerCase() === 'time') return [];
  
  // 1. Check if contains standard batch pattern like FA24-BCS-A, SP23-FSN
  const batchMatches = s.match(/(?:FA|SP)\d{2}-[A-Z0-9]+(?:-[A-Z0-9]+)*/gi);
  if (batchMatches && batchMatches.length > 0) {
    // If multiple batches fused like FA25-BME-FA24-BME-FA22-BEE
    let results = [];
    for (let bm of batchMatches) {
      // Split on secondary FA/SP if chained
      const sub = bm.split(/(?=(?:FA|SP)\d{2}-)/i).map(b => b.replace(/^[-/, ]+|[-/, ]+$/g, '').trim()).filter(Boolean);
      results.push(...sub);
    }
    return results;
  }
  
  // 2. Fallback split on slash or comma
  return s.split(/[\/,]+/).map(b => b.trim()).filter(Boolean);
}

function cleanRoomName(raw) {
  if (!raw) return "";
  let r = String(raw).trim();
  // Strip capacity e.g. "A - 3 (42)" -> "A-3", "B 4 (20)" -> "B-4", "C 1.1 (49)" -> "C-1.1", " D1 (42)" -> "D-1"
  r = r.replace(/\s*\(\d+\)\s*/g, '');
  r = r.replace(/\s*-\s*/g, '-');
  r = r.replace(/\s+/g, '-');
  return r.trim();
}

function formatExamTime(raw) {
  if (!raw) return "";
  let t = String(raw).trim();
  // "0900-1200" -> "09:00 AM - 12:00 PM"
  // "0100-0400" -> "01:00 PM - 04:00 PM"
  // "0900-1030" -> "09:00 AM - 10:30 AM"
  // "1100-1230" -> "11:00 AM - 12:30 PM"
  // "0100-0230" -> "01:00 PM - 02:30 PM"
  // "0300-0430" -> "03:00 PM - 04:30 PM"
  const m = t.match(/^(\d{2})(\d{2})\s*-\s*(\d{2})(\d{2})$/);
  if (m) {
    let sh = parseInt(m[1], 10);
    let sm = m[2];
    let eh = parseInt(m[3], 10);
    let em = m[4];
    
    // In university exams, slots are 0900, 1100, 0100 (13:00), 0300 (15:00)
    let sPeriod = (sh >= 8 && sh <= 11) ? "AM" : "PM";
    let ePeriod = (eh >= 8 && eh <= 11) ? "AM" : "PM";
    if (sh === 1 || sh === 2 || sh === 3 || sh === 4 || sh === 5) sPeriod = "PM";
    if (eh === 1 || eh === 2 || eh === 3 || eh === 4 || eh === 5 || eh === 12) ePeriod = "PM";
    
    let sDisp = sh.toString().padStart(2, '0');
    let eDisp = eh.toString().padStart(2, '0');
    return `${sDisp}:${sm} ${sPeriod} - ${eDisp}:${em} ${ePeriod}`;
  }
  return t;
}

function parseExcelQuadMatrix(filePath) {
  console.log(`\n======================================================`);
  console.log(`TESTING EXCEL QUAD-MATRIX PARSER: ${path.basename(filePath)}`);
  console.log(`======================================================`);
  
  const buffer = fs.readFileSync(filePath);
  const workbook = XLSX.read(buffer, { type: 'buffer' });
  const sheet = workbook.Sheets[workbook.SheetNames[0]];
  const rows = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: null });
  
  // Find header row (row with Date / Time / Room names)
  let headerRowIdx = -1;
  for (let rIdx = 0; rIdx < Math.min(rows.length, 10); rIdx++) {
    const row = rows[rIdx] || [];
    const dateCount = row.filter(c => c && String(c).trim().toLowerCase() === 'date').length;
    if (dateCount >= 2) {
      headerRowIdx = rIdx;
      break;
    }
  }
  if (headerRowIdx === -1) headerRowIdx = 2;
  
  const headerRow = rows[headerRowIdx] || [];
  
  // Identify all 4 campus blocks
  const blocks = [];
  for (let c = 0; c < headerRow.length; c++) {
    const val = headerRow[c];
    if (val && String(val).trim().toLowerCase() === 'date') {
      const dateCol = c;
      const timeCol = c + 1;
      blocks.push({ dateCol, timeCol, startCol: c + 2, endCol: headerRow.length });
    }
  }
  for (let i = 0; i < blocks.length; i++) {
    if (i + 1 < blocks.length) {
      blocks[i].endCol = blocks[i + 1].dateCol;
    }
  }
  
  console.log(`Detected ${blocks.length} Horizontal Campus Blocks:`);
  blocks.forEach((b, idx) => {
    const rooms = [];
    for (let c = b.startCol; c < b.endCol; c++) {
      if (headerRow[c]) rooms.push({ colIdx: c, name: cleanRoomName(headerRow[c]) });
    }
    b.rooms = rooms;
    console.log(`  Block ${idx + 1} (Cols ${b.dateCol}..${b.endCol-1}): ${rooms.length} rooms -> ${rooms.map(r => r.name).join(', ')}`);
  });
  
  const parsedExams = [];
  const currentDates = Array(blocks.length).fill(null);
  
  let r = headerRowIdx + 1;
  while (r < rows.length) {
    const row = rows[r] || [];
    const nextRow = rows[r + 1] || [];
    
    // Check if this row is completely empty
    if (!row.some(Boolean) && !nextRow.some(Boolean)) {
      r += 1;
      continue;
    }
    
    // Process each block independently
    for (let bIdx = 0; bIdx < blocks.length; bIdx++) {
      const b = blocks[bIdx];
      const rawDate = row[b.dateCol];
      const rawTime = row[b.timeCol];
      
      const dateStr = rawDate ? String(rawDate).trim() : "";
      const timeStr = rawTime ? String(rawTime).trim() : "";
      
      if (dateStr && dateStr.toLowerCase() !== 'date') {
        currentDates[bIdx] = dateStr;
      }
      
      const activeDate = currentDates[bIdx] || currentDates[0] || "Unknown Date";
      const activeTime = formatExamTime(timeStr);
      
      if (!timeStr || timeStr.toLowerCase() === 'time') {
        continue;
      }
      
      for (const room of b.rooms) {
        const batchCell = row[room.colIdx];
        const subjectCell = nextRow[room.colIdx];
        
        if (batchCell === null || batchCell === undefined) continue;
        const bStr = String(batchCell).trim();
        if (bStr === '' || bStr.toLowerCase() === 'none' || bStr.toLowerCase() === 'date' || bStr.toLowerCase() === 'time') continue;
        
        const sStr = subjectCell !== null && subjectCell !== undefined ? String(subjectCell).trim() : "";
        
        const batches = splitCombinedBatches(bStr);
        if (batches.length === 0) continue;
        
        for (let batch of batches) {
          parsedExams.push({
            date: activeDate,
            time: activeTime,
            room: room.name,
            batch: batch,
            subject: sStr || "EXAM"
          });
        }
      }
    }
    r += 2;
  }
  
  console.log(`\nSuccessfully Parsed ${parsedExams.length} Individual Student Cohort Exam Records!`);
  console.log(`Sample Records:`);
  console.log(parsedExams.slice(0, 5));
  
  const uniqueBatches = new Set(parsedExams.map(e => e.batch));
  console.log(`Unique Batches Count: ${uniqueBatches.size} (e.g. ${Array.from(uniqueBatches).slice(0, 8).join(', ')})`);
  return parsedExams;
}

parseExcelQuadMatrix("assets/documents/Date_Sheet_FINAL_Term_Exam_FALL_2025 - Version I - 08-12-25.xlsx");
parseExcelQuadMatrix("assets/documents/Version I of Date Sheet Mid Term Exam SPRING 2026 25-03-2026.xlsx");

File: tools/test_excel_quad_parser.py
import openpyxl
import os
import re

def split_combined_batches(raw):
    if not raw: return []
    s = str(raw).strip()
    if s == '' or s.lower() == 'none' or s.lower() == 'date' or s.lower() == 'time': return []
    
    # Check if contains standard batch pattern like FA24-BCS-A, SP23-FSN
    batch_matches = re.findall(r'(?:FA|SP)\d{2}-[A-Z0-9]+(?:-[A-Z0-9]+)*', s, re.I)
    if batch_matches:
        results = []
        for bm in batch_matches:
            # Split on secondary FA/SP if chained e.g. FA25-BME-FA24-BME-FA22-BEE
            sub = re.split(r'(?=(?:FA|SP)\d{2}-)', bm, flags=re.I)
            for b in sub:
                cleaned = re.sub(r'^[-/, ]+|[-/, ]+$', '', b).strip()
                if cleaned: results.append(cleaned)
        return results
    return [b.strip() for b in re.split(r'[/,]+', s) if b.strip()]

def clean_room_name(raw):
    if not raw: return ""
    r = str(raw).strip()
    r = re.sub(r'\s*\(\d+\)\s*', '', r)
    r = re.sub(r'\s*-\s*', '-', r)
    r = re.sub(r'\s+', '-', r)
    return r.strip()

def format_exam_time(raw):
    if not raw: return ""
    t = str(raw).strip()
    m = re.match(r'^(\d{2})(\d{2})\s*-\s*(\d{2})(\d{2})$', t)
    if m:
        sh, sm, eh, em = int(m.group(1)), m.group(2), int(m.group(3)), m.group(4)
        s_period = "AM" if 8 <= sh <= 11 else "PM"
        e_period = "AM" if 8 <= eh <= 11 else "PM"
        if sh in [1, 2, 3, 4, 5]: s_period = "PM"
        if eh in [1, 2, 3, 4, 5, 12]: e_period = "PM"
        return f"{sh:02d}:{sm} {s_period} - {eh:02d}:{em} {e_period}"
    return t

def parse_excel_quad(file_path):
    print("=" * 80)
    print(f"VERIFYING QUAD-MATRIX PARSER ON: {os.path.basename(file_path)}")
    print("=" * 80)
    
    wb = openpyxl.load_workbook(file_path, data_only=True)
    sheet = wb.active
    rows = list(sheet.iter_rows(values_only=True))
    
    header_row_idx = -1
    for r_idx in range(min(len(rows), 10)):
        row = rows[r_idx]
        dates = [c for c in row if c and str(c).strip().lower() == 'date']
        if len(dates) >= 2:
            header_row_idx = r_idx
            break
    if header_row_idx == -1: header_row_idx = 2
    
    header_row = rows[header_row_idx]
    
    blocks = []
    for c_idx, val in enumerate(header_row):
        if val and str(val).strip().lower() == 'date':
            blocks.append({
                'date_col': c_idx,
                'time_col': c_idx + 1,
                'start_col': c_idx + 2,
                'end_col': len(header_row)
            })
    for i in range(len(blocks)):
        if i + 1 < len(blocks):
            blocks[i]['end_col'] = blocks[i+1]['date_col']
            
    for idx, b in enumerate(blocks):
        rooms = []
        for c in range(b['start_col'], b['end_col']):
            if header_row[c]:
                rooms.append({'col_idx': c, 'name': clean_room_name(header_row[c])})
        b['rooms'] = rooms
        print(f"  Block {idx+1} (Cols {b['date_col']}..{b['end_col']-1}): {len(rooms)} rooms -> {[r['name'] for r in rooms[:4]]}")
        
    parsed_exams = []
    current_dates = [None] * len(blocks)
    
    r = header_row_idx + 1
    while r < len(rows):
        row = rows[r]
        next_row = rows[r + 1] if r + 1 < len(rows) else [None] * len(row)
        
        if not any(row) and not any(next_row):
            r += 1
            continue
            
        for b_idx, b in enumerate(blocks):
            raw_date = row[b['date_col']]
            raw_time = row[b['time_col']]
            
            date_str = str(raw_date).strip() if raw_date is not None else ""
            time_str = str(raw_time).strip() if raw_time is not None else ""
            
            if date_str and date_str.lower() != 'date':
                current_dates[b_idx] = date_str
                
            active_date = current_dates[b_idx] or current_dates[0] or "Unknown Date"
            active_time = format_exam_time(time_str)
            
            if not time_str or time_str.lower() == 'time':
                continue
                
            for room in b['rooms']:
                b_cell = row[room['col_idx']]
                s_cell = next_row[room['col_idx']]
                
                if b_cell is None: continue
                b_str = str(b_cell).strip()
                if not b_str or b_str.lower() in ['none', 'date', 'time']: continue
                
                s_str = str(s_cell).strip() if s_cell is not None else ""
                batches = split_combined_batches(b_str)
                if not batches: continue
                
                for batch in batches:
                    parsed_exams.append({
                        'date': active_date,
                        'time': active_time,
                        'room': room['name'],
                        'batch': batch,
                        'subject': s_str or "EXAM"
                    })
        r += 2
        
    print(f"\nExtracted {len(parsed_exams)} Individual Cohort Exam Slots!")
    batches_set = set(e['batch'] for e in parsed_exams)
    print(f"Total Unique Batches: {len(batches_set)} (e.g. {list(batches_set)[:6]})")
    print("Sample Output Records:")
    for item in parsed_exams[:4]:
        print(f"  {item}")

if __name__ == '__main__':
    parse_excel_quad("assets/documents/Date_Sheet_FINAL_Term_Exam_FALL_2025 - Version I - 08-12-25.xlsx")
    parse_excel_quad("assets/documents/Version I of Date Sheet Mid Term Exam SPRING 2026 25-03-2026.xlsx")
